File: src/python_examples/cgalpy_examples/Arrangement_on_surface_2/aos2_visual_buffer_raster.cpp
// Author(s): Utkarsh Khajuria <utkarshkhajuria55@gmail.com>

#include "arr_exact_construction_segments.h"

#include <CGAL/draw_arrangement_2.h>
#include <CGAL/Graphics_scene.h>

#include <cmath>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {

constexpr int width = 192;
constexpr int height = 160;
constexpr int margin = 8;
constexpr double x_min = 0.0;
constexpr double x_max = 6.0;
constexpr double y_min = 0.0;
constexpr double y_max = 5.0;

std::pair<int, int> map_point(double x, double y) {
  if (x < x_min || x > x_max || y < y_min || y > y_max) {
    throw std::runtime_error("point outside fixed viewport");
  }
  const int px = margin + static_cast<int>(std::floor(
    ((x - x_min) / (x_max - x_min))
      * static_cast<double>(width - 1 - 2 * margin) + 0.5));
  const int py = height - 1 - margin - static_cast<int>(std::floor(
    ((y - y_min) / (y_max - y_min))
      * static_cast<double>(height - 1 - 2 * margin) + 0.5));
  return {px, py};
}

void set_black(std::vector<unsigned char>& image, int x, int y) {
  if (x < 0 || x >= width || y < 0 || y >= height) {
    throw std::runtime_error("pixel outside raster");
  }
  const std::size_t offset =
    static_cast<std::size_t>((y * width + x) * 3);
  image[offset] = 0;
  image[offset + 1] = 0;
  image[offset + 2] = 0;
}

void draw_line(std::vector<unsigned char>& image,
               int x0, int y0, int x1, int y1) {
  const int dx = std::abs(x1 - x0);
  const int sx = x0 < x1 ? 1 : -1;
  const int dy = -std::abs(y1 - y0);
  const int sy = y0 < y1 ? 1 : -1;
  int error = dx + dy;

  while (true) {
    set_black(image, x0, y0);
    if (x0 == x1 && y0 == y1) {
      break;
    }
    const int twice_error = 2 * error;
    if (twice_error >= dy) {
      error += dy;
      x0 += sx;
    }
    if (twice_error <= dx) {
      error += dx;
      y0 += sy;
    }
  }
}

void write_ppm(const std::string& path,
               const std::vector<unsigned char>& image) {
  std::ofstream output(path, std::ios::binary);
  if (!output) {
    throw std::runtime_error("unable to open raster output");
  }
  output << "P6\n" << width << ' ' << height << "\n255\n";
  output.write(
    reinterpret_cast<const char*>(image.data()),
    static_cast<std::streamsize>(image.size()));
  if (!output) {
    throw std::runtime_error("failed while writing raster output");
  }
}

} // namespace

int main(int argc, char** argv) {
  try {
    if (argc != 2) {
      std::cerr << "usage: aos2_visual_buffer_raster <output.ppm>\n";
      return 2;
    }

    Segment segments[] = {
      Segment(Point(1, 0), Point(2, 4)),
      Segment(Point(5, 0), Point(5, 5)),
      Segment(Point(1, 0), Point(5, 3)),
      Segment(Point(0, 2), Point(6, 0)),
      Segment(Point(3, 0), Point(5, 5))
    };

    Arrangement arrangement;
    insert(
      arrangement,
      segments,
      segments + sizeof(segments) / sizeof(Segment));

    CGAL::Graphics_scene scene;
    CGAL::add_to_graphics_scene(arrangement, scene);

    const auto& buffer =
      scene.get_array_of_index(CGAL::Graphics_scene::POS_SEGMENTS);
    if (buffer.size() % 6 != 0) {
      throw std::runtime_error(
        "POS_SEGMENTS float count is not divisible by 6");
    }

    std::vector<unsigned char> image(
      static_cast<std::size_t>(width * height * 3),
      static_cast<unsigned char>(255));

    for (std::size_t offset = 0; offset < buffer.size(); offset += 6) {
      const double x0 = buffer[offset];
      const double y0 = buffer[offset + 1];
      const double z0 = buffer[offset + 2];
      const double x1 = buffer[offset + 3];
      const double y1 = buffer[offset + 4];
      const double z1 = buffer[offset + 5];

      if (std::abs(z0) > 1e-6 || std::abs(z1) > 1e-6) {
        throw std::runtime_error(
          "expected the Aos2 scene to lie in the XY plane");
      }

      const auto p0 = map_point(x0, y0);
      const auto p1 = map_point(x1, y1);
      draw_line(image, p0.first, p0.second, p1.first, p1.second);
    }

    write_ppm(argv[1], image);
    std::cout << "raster_width " << width << '\n';
    std::cout << "raster_height " << height << '\n';
    std::cout << "raster_segments " << buffer.size() / 6 << '\n';
    return 0;
  }
  catch (const std::exception& error) {
    std::cerr << error.what() << '\n';
    return 1;
  }
}

File: src/python_examples/cgalpy_examples/Arrangement_on_surface_2/aos2_visual_buffer_raster.py
#!/usr/bin/python
# Author(s): Utkarsh Khajuria <utkarshkhajuria55@gmail.com>

import importlib
import math
import sys
from pathlib import Path

WIDTH = 192
HEIGHT = 160
MARGIN = 8
X_MIN = 0.0
X_MAX = 6.0
Y_MIN = 0.0
Y_MAX = 5.0


def map_point(x, y):
    if not (X_MIN <= x <= X_MAX and Y_MIN <= y <= Y_MAX):
        raise ValueError(f"point outside fixed viewport: ({x}, {y})")
    px = MARGIN + int(
        math.floor(
            ((x - X_MIN) / (X_MAX - X_MIN))
            * (WIDTH - 1 - 2 * MARGIN)
            + 0.5
        )
    )
    py = HEIGHT - 1 - MARGIN - int(
        math.floor(
            ((y - Y_MIN) / (Y_MAX - Y_MIN))
            * (HEIGHT - 1 - 2 * MARGIN)
            + 0.5
        )
    )
    return px, py


def set_black(image, x, y):
    if not (0 <= x < WIDTH and 0 <= y < HEIGHT):
        raise ValueError(f"pixel outside raster: ({x}, {y})")
    offset = (y * WIDTH + x) * 3
    image[offset] = 0
    image[offset + 1] = 0
    image[offset + 2] = 0


def draw_line(image, x0, y0, x1, y1):
    dx = abs(x1 - x0)
    sx = 1 if x0 < x1 else -1
    dy = -abs(y1 - y0)
    sy = 1 if y0 < y1 else -1
    error = dx + dy

    while True:
        set_black(image, x0, y0)
        if x0 == x1 and y0 == y1:
            break
        twice_error = 2 * error
        if twice_error >= dy:
            error += dy
            x0 += sx
        if twice_error <= dx:
            error += dx
            y0 += sy


def write_ppm(path, image):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(
        f"P6\n{WIDTH} {HEIGHT}\n255\n".encode("ascii")
        + bytes(image)
    )


if len(sys.argv) != 3:
    raise SystemExit(
        "usage: aos2_visual_buffer_raster.py "
        "<binding-library> <output.ppm>"
    )

library = sys.argv[1]
output = Path(sys.argv[2])

CGALPY = importlib.import_module(library)
Aos2 = CGALPY.Aos2
Bvr = CGALPY.Bvr

Arrangement = Aos2.Arrangement_2
Point = Arrangement.Geometry_traits_2.Point_2
Segment = Arrangement.Geometry_traits_2.X_monotone_curve_2

segments = [
    Segment(Point(1, 0), Point(2, 4)),
    Segment(Point(5, 0), Point(5, 5)),
    Segment(Point(1, 0), Point(5, 3)),
    Segment(Point(0, 2), Point(6, 0)),
    Segment(Point(3, 0), Point(5, 5)),
]

arrangement = Arrangement()
Aos2.insert(arrangement, segments)

scene = Bvr.Graphics_scene()
Bvr.add_to_graphics_scene(arrangement, scene)

buffer = scene.buffer(Bvr.POS_SEGMENTS)
if len(buffer) % 6 != 0:
    raise RuntimeError(
        f"POS_SEGMENTS float count is not divisible by 6: {len(buffer)}"
    )

image = bytearray([255]) * (WIDTH * HEIGHT * 3)

for offset in range(0, len(buffer), 6):
    x0, y0, z0, x1, y1, z1 = buffer[offset:offset + 6]
    if abs(z0) > 1e-6 or abs(z1) > 1e-6:
        raise RuntimeError("expected the Aos2 scene to lie in the XY plane")
    p0 = map_point(float(x0), float(y0))
    p1 = map_point(float(x1), float(y1))
    draw_line(image, p0[0], p0[1], p1[0], p1[1])

write_ppm(output, image)
print("raster_width", WIDTH)
print("raster_height", HEIGHT)
print("raster_segments", len(buffer) // 6)
